core/break.lua

```diff
@@ -75,30 +74,42 @@ function lineBreak:trimGlue() -- 842
 end
 
 function lineBreak:setupLineLengths() -- 874
-  self.parShape = param("parShape")
+  -- NOTE FOR DEVELOPERS: the parShape case is not handled here and the
+  -- paragraph shapes are intended to be provided by extending this method
+  -- (and possibly other typesetter methods) outside of the default
+  -- implementation here.
+  --
+  -- It expect a method parShape(self, line) returning a table
+  --    { width=..., left=..., right=... }
+  -- defining the width, left indent and right indent for the given line
+  -- (where obviously, left + right + width = self.hsize, so returning the
+  -- two indents could have been avoided, but it feels more readable and natural
+  -- this way, and easier afterwards back into the typesetter).
+  -- The width must be a SILE.measurement()'s, and the indents absolute
+  -- numbers.
+  --
+  -- Using a Lua function rather than a fixed array allows one to perform
+  -- computations, but remember the linebreak methods will be called multiple
+  -- times to try good breaks, so it is strongly recommended to "memoize"
+  -- values for lines already computed.
+  --
+  -- TeX wizards shall also note that this is slighty different from
+  -- Knuth's definition "nline l1 i1 l2 i2 ... lN iN".
+  self.parShape = nil
```

Comment:
I must say that this way of doing things has a very strong code smell. This `parShape` should not be here at all as it's never used by the function and you show that the only way to use it is to rewrite the whole function…it would be better for there to be a `lineBreak:parShape` method, which by default is a noop (rather than being tested for `nil`), but which can be rewritten by user scripts. This would also mean you don't need to save and restore `lineBreak:setupLineLengths`, only `parShape`; and indeed if the noop function is simple to rewrite not even that.

Reply:
You are right (well rather than a noop, just spit out the h.size) + a setting to enable/disable it would allow to fix the case of folios and headers (as we do for the hangs). Working on it.

Reply:
Done, and indeed clearer IMHO